src/check.rs

```diff
@@ -61,16 +86,81 @@ pub fn is_available(url: &Url, ctx: &CheckContext) -> Result<(), CheckError> {
+        Ok(())
+    } else {
+        // Rust documentation uses `#n-m` fragments and JavaScript to highlight
+        // a range of lines in HTML of source code, an element with `id`
+        // attribute of (literal) "#n-m" will not exist, but elements with
+        // `id`s n through m should, this parses the ranged n-m anchor and
+        // checks if elements with `id`s n through m do exist
+        match RUST_LINE_HIGLIGHT_RX.captures(fragment) {
+            Some(capture) => match (capture.name("start"), capture.name("end")) {
+                (Some(start_str), Some(end_str)) => {
+                    let start = start_str.as_str().parse::<i32>().unwrap();
+                    let end = end_str.as_str().parse::<i32>().unwrap();
+                    let missing = (start..=end)
+                        .map(|i| i.to_string())
+                        .filter(|i| !fragments.contains(i))
+                        .collect::<Vec<String>>();
+                    if !missing.is_empty() {
+                        Err(CheckError::AnchorRange(
+                            url.clone(),
+                            fragment.to_string(),
+                            missing,
+                        ))
+                    } else {
+                        Ok(())
+                    }
+                }
+                // Should not be reached as we expect regex to match and capture
+                _ => Err(CheckError::Anchor(url.clone(), fragment.to_string())),
+            },
+            None => Err(CheckError::Anchor(url.clone(), fragment.to_string())),
+        }
+    }
+}
+
 /// Check a URL with the "file" scheme for availability. Returns `false` if it is unavailable.
 fn check_file_url(url: &Url, _ctx: &CheckContext) -> Result<(), CheckError> {
     let path = url.to_file_path().unwrap();
 
-    if path.is_file() || path.join("index.html").is_file() {
-        debug!("Linked file at path {} does exist.", path.display());
-        Ok(())
+    // determine the full path by looking if the path points to a directory,
+    // and if so append `index.html`, this is needed as we'll try to read
+    // the file, so `expanded_path` should point to a file not a directory
+    let expanded_path = if path.is_file() {
+        path
+    } else if path.is_dir() {
+        path.join("index.html")
     } else {
         debug!("Linked file at path {} does not exist!", path.display());
-        Err(CheckError::File(path))
+        return Err(CheckError::File(path));
+    };
+
+    debug!(
+        "Linked file at path {} does exist.",
+        expanded_path.display()
+    );
+
+    // the URL might contain a fragment, in that case we need to check if
+    // the fragment exists in the file
+    match url.fragment() {
+        Some(fragment) => check_file_fragment(url, &expanded_path, fragment),
+        None => Ok(()),
+    }
+}
+
+cached_key_result! {
+    CHECK_FILE: UnboundCache<(String, String), ()> = UnboundCache::new();
```

Comment:
Hmm, I tested on `stm32h7xx_hal` and this went up to about 300 MB (stm23h7 has about 100k HTML files in total). Do you think that's worth switching to a bounded cache? Normally the memory usage is a tiny fraction of that, like 50 KB or so.

Reply:
I've opted on using SizedCache with 100 entries.